contracts/AelinPool.sol

```diff
@@ -209,49 +212,99 @@ contract AelinPool is AelinERC20, MinimalProxyFactory, IAelinPool {
 
+        //Values re-declared each loop
+        NftPurchaseList memory nftPurchaseList;
+        address collectionAddress;
+        uint256[] memory tokenIds;
+        uint256 tokenIdsLength;
+        NftCollectionRules memory nftCollectionRules;
+
+        //The running total for 721 tokens
         uint256 maxPurchaseTokenAmount;
 
-        for (uint256 i; i < _nftPurchaseList.length; ++i) {
-            NftPurchaseList memory nftPurchaseList = _nftPurchaseList[i];
-            address collectionAddress = nftPurchaseList.collectionAddress;
-            uint256[] memory tokenIds = nftPurchaseList.tokenIds;
+        //Iterate over the collections
+        for (uint256 i; i < nftPurchaseListLength; ++i) {
+            nftPurchaseList = _nftPurchaseList[i];
+            collectionAddress = nftPurchaseList.collectionAddress;
+            tokenIds = nftPurchaseList.tokenIds;
+            tokenIdsLength = tokenIds.length;
+            nftCollectionRules = nftCollectionDetails[collectionAddress];
 
             require(collectionAddress != address(0), "collection should not be null");
-
-            NftCollectionRules memory nftCollectionRules = nftCollectionDetails[collectionAddress];
             require(nftCollectionRules.collectionAddress == collectionAddress, "collection not in the pool");
 
-            if (nftCollectionRules.purchaseAmount > 0) {
-                maxPurchaseTokenAmount += nftCollectionRules.purchaseAmount * tokenIds.length;
-            }
-
-            if (nftCollectionRules.purchaseAmount == 0) {
-                maxPurchaseTokenAmount = _purchaseTokenAmount;
-            }
+            //Iterate over the token ids
+            for (uint256 j; j < tokenIdsLength; ++j) {
+                if (NftCheck.supports721(collectionAddress)) {
+                    require(IERC721(collectionAddress).ownerOf(tokenIds[j]) == msg.sender, "has to be the token owner");
+                    // If there are no ranges then no need to check whether token Id is within them
+                    // Or whether there are any rangeAmounts
+                    if (nftCollectionRules.idRanges.length > 0) {
+                        (bool isTokenIdInRange, uint256 rangeAmountForTokenId) = getRangeData(
+                            tokenIds[j],
+                            nftCollectionRules.idRanges
+                        );
+                        require(isTokenIdInRange, "tokenId not in range");
+
+                        //if there's a range amount for this token id, then set that as the max for its element
+                        if (rangeAmountForTokenId != 0) {
+                            maxPurchaseTokenAmount = incrementMaxPurchaseTokenAmount(
+                                maxPurchaseTokenAmount,
+                                rangeAmountForTokenId
+                            );
+                        } else {
+                            //Otherwise defer to purchaseAmount
```

Comment:
this block and the following are identical. Maybe we can put these in a function? Something like `getDefaultMaxPurchaseAmount`. I will let you think of a name you like